### PR TITLE
[FLINK-10135] The JobManager does not report the cluster-level metrics

### PR DESCRIPTION
**Pull request by yanghua**
## What is the purpose of the change

*This pull request fixed job manager does not report the cluster-level metrics*

## Brief change log

  - *Fixed job manager does not report the cluster-level metrics*

## Verifying this change

This change is already covered by existing tests, such as *TestingRestfulGateway#requestMultipleJobsOverview*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by yanghua** [comment]
@GJL @tillrohrmann @dawidwys  can you review this PR?

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java [hunk 1]
```diff
@@ -111,6 +112,14 @@
 	 */
 	CompletableFuture<ClusterOverview> requestClusterOverview(@RpcTimeout Time timeout);
 
+	/**
+	 * Requests the all the jobs overview.
+	 *
+	 * @param timeout for the asynchronous operation
+	 * @return Future containing the status overview
+	 */
+	CompletableFuture<JobsOverview> requestMultipleJobsOverview(Time timeout);
```
Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java]: Add `@RpcTimeout`

File: flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingDispatcherGateway.java [hunk 2]
```diff
@@ -70,25 +71,26 @@ public TestingDispatcherGateway() {
...
-			Supplier<CompletableFuture<Collection<String>>> requestMetricQueryServicePathsSupplier,
-			Supplier<CompletableFuture<Collection<Tuple2<ResourceID, String>>>> requestTaskManagerMetricQueryServicePathsSupplier,
-			BiFunction<JobID, JobVertexID, CompletableFuture<OperatorBackPressureStatsResponse>> requestOperatorBackPressureStatsFunction,
-			BiFunction<JobID, String, CompletableFuture<String>> triggerSavepointFunction,
-			Function<JobGraph, CompletableFuture<Acknowledge>> submitFunction,
-			Supplier<CompletableFuture<Collection<JobID>>> listFunction,
-			int blobServerPort,
-			DispatcherId fencingToken,
-			Function<JobID, CompletableFuture<ArchivedExecutionGraph>> requestArchivedJobFunction) {
+		String address,
```
Comment (StefanRRichter) [inline flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingDispatcherGateway.java]: Unrelated formatting changes.

File: flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingRestfulGateway.java [hunk 3]
```diff
@@ -100,24 +104,26 @@ public TestingRestfulGateway() {
...
-			Function<JobID, CompletableFuture<? extends AccessExecutionGraph>> requestJobFunction,
-			Function<JobID, CompletableFuture<JobResult>> requestJobResultFunction,
-			Function<JobID, CompletableFuture<JobStatus>> requestJobStatusFunction,
-			Supplier<CompletableFuture<MultipleJobsDetails>> requestMultipleJobDetailsSupplier,
-			Supplier<CompletableFuture<ClusterOverview>> requestClusterOverviewSupplier,
-			Supplier<CompletableFuture<Collection<String>>> requestMetricQueryServicePathsSupplier,
-			Supplier<CompletableFuture<Collection<Tuple2<ResourceID, String>>>> requestTaskManagerMetricQueryServicePathsSupplier,
-			BiFunction<JobID, JobVertexID, CompletableFuture<OperatorBackPressureStatsResponse>> requestOperatorBackPressureStatsFunction,
-			BiFunction<JobID, String, CompletableFuture<String>> triggerSavepointFunction) {
+		String address,
```
Comment (StefanRRichter) [inline flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingRestfulGateway.java]: Unrelated formatting changes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java [hunk 4]
```diff
@@ -111,6 +112,14 @@
 	 */
 	CompletableFuture<ClusterOverview> requestClusterOverview(@RpcTimeout Time timeout);
 
+	/**
+	 * Requests the all the jobs overview.
```
[the remark this answers is not in the document]
Comment (yanghua) [reply flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java]: Because I saw the implementation of `requestClusterOverview ` combines the completed jobs' overview, but this method would not contains so I refer to `requestMultipleJobDetails`. All right, it seems there is no problem about the "all". Will change it.

**Comment by yanghua** [comment]
@StefanRRichter thanks for your review suggestion, I have refactored the PR. I have tried to test `requestOverviewForAllJobs` in `DispatcherTest`, but always throw akka timeout : 

```
Caused by: akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka://flink/user/dispatcher_testJobRecovery#-798214439]] after [10000 ms]. Sender[null] sent message of type "org.apache.flink.runtime.rpc.messages.LocalFencedMessage".
```

I replace the method with `requestClusterOverview`, the result is the same exception. So I do not know how to test it. It seems there is not a case can be refer to. So can you give a little hint? thanks.

**Comment by yanghua** [comment]
@StefanRRichter I have tried to add a test case for reporting job manager metrics, can you review this?

**Comment by yanghua** [comment]
@zentol updated the PR.

**Comment by yanghua** [comment]
@StefanRRichter can you review it?

File: flink-runtime/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.junit.Assert.fail;
+
+/**
+ * The test class for metrics report by job manager.
+ */
+public class JobManagerMetricsTest {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsTest.java]: This isn't testing anything.

What we want to ensure is that these metrics don't disappear again, and optionally that the values are correct. For this you have to actually start runtime components, which could be done with a `MiniCluster` (that is also configured to use the TestReporter) and have the reporter search for the desired metrics.
The `SystemResourcesMetricsITCase` may be a useful reference

Comment (yanghua) [reply flink-runtime/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsTest.java]: OK. Thanks, I will try it.

**Comment by yanghua** [comment]
@zentol I have changed the test case, please review it again, thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java [hunk 6]
```diff
@@ -909,6 +926,70 @@ private void clearDispatcherState() {
...
+			} catch (ExecutionException e) {
+				log.error("Request resource overview occurs exception.", e);
+			}
+
+			return 0L;
+		});
+
+		jobManagerMetricGroup.gauge("numRunningJobs", () -> {
+			try {
+				return (long) requestOverviewForAllJobs(defaultRequestTimeout)
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java]: This results in a wave of RPC calls to every single JM. Would it not be feasible to simple count the number of job IDs contained in `jobManagerRunnerFutures`?

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+	@Before
+	public void setUp() throws Exception {
+		invokeLatch = new CountDownLatch(1);
+
+		Assume.assumeTrue(
+			"ClusterClient is not an instance of MiniClusterClient",
+			MINI_CLUSTER_RESOURCE.getClusterClient() instanceof MiniClusterClient);
+
+		clusterClient = (MiniClusterClient) MINI_CLUSTER_RESOURCE.getClusterClient();
+		clusterClient.setDetached(true);
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: You could simplify this section by simple defining a job that has a blocking UDF and executing it in a separate thread.

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+				assertEquals(1L, entry.getKey().getValue());
+			} else if (entry.getValue().contains(NUM_RUNNING_JOBS)) {
+				assertEquals(1L, entry.getKey().getValue());
+			}
+		}
+	}
+
+	private static Configuration getConfiguration() {
+		Configuration configuration = new Configuration();
+		configuration.setString(REPORTERS_LIST, "test_reporter");
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: this configuration setting is not necessary anymore, reporter names are automatically detected

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+				assertThat(jobStatus.isGloballyTerminalState(), equalTo(false));
+				if (jobStatus == JobStatus.RUNNING) {
+					return;
+				}
+			} catch (ExecutionException ignored) {
+				// JobManagerRunner is not yet registered in Dispatcher
+			}
+			Thread.sleep(1000);
+		}
+		throw new AssertionError("Job did not become running within timeout.");
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: use `fail()` instead

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	private static Configuration getConfiguration() {
+		Configuration configuration = new Configuration();
+		configuration.setString(REPORTERS_LIST, "test_reporter");
+		configuration.setString("metrics.reporter.test_reporter.class", TestReporter.class.getName());
+		return configuration;
+	}
+
+	private void waitForJob() throws Exception {
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: This method should be called `waitForJobRunning()`; however this entire logic can be implemented more nicely with a [BlockerSync](https://github.com/apache/flink/blob/master/flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/BlockerSync.java) anyway. The UDF would wait until the test thread calls `block`, and then you reverse this for unblocking the thread.

This method could then be removed.

**Comment by yanghua** [comment]
@zentol Thanks for your suggestion, I have updated the implementation, please review it again.

**Comment by zentol** [comment]
This looks a lot better, the majority of my comments are for minor things, but we've now run into a concurrency issue. All `RpcEndpoints` (i.e. DIspatcher, ResourceManager, etc) assume to run in a single thread (the "main-thread"), but metrics are accessed by another one. We have to either make sure that all operations are safe in this scenario, or _somehow_ (@tillrohrmann?) defer the access to the main-thread.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java [hunk 11]
```diff
@@ -734,6 +744,18 @@ public void requestHeartbeat(ResourceID resourceID, Void payload) {
 		}
 	}
 
+	private void registerSlotAndTaskExecutorMetrics() {
+		jobManagerMetricGroup.gauge(
+			TASK_SLOTS_AVAILABLE_METRIC_NAME,
+			() -> (long) slotManager.getNumberFreeSlots());
```
[the remark this answers is not in the document]
Comment (yanghua) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: @zentol There may be problems with read inconsistencies here. But I think the impact is small:

* If RM and Dispatcher are single-threaded, there will be no concurrent modification problems here;
* The indicator is periodically extracted (inconsistent is only a short period of one cycle) and its value does not play any decision role in Flink (should be tolerable?)
* Considering the frequency of metric extraction, if we provide related synchronization mechanisms, performance will be reduced and complexity will be increased.

What do you think?

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: I think it should be okish to call these methods from a different thread. All of them simply return the size of a `HashMap`. In the worst case, we would report a little bit stale data. The main danger I see is that the implementations of `SlotManager` might change, which makes this no longer hold true.

If we wanted to make it "thread-safe", then we would need to make the access call inside of the `MainThreadExecutor`:
```
jobManagerMetricGroup.gauge(
	MetricNames.TASK_SLOTS_AVAILABLE,
	() -> callAsync(() -> slotManager.getNumberFreeSlots(), timeout).get());
```

However, this might be an overkill at the moment.

File: flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerTest.java [hunk 12]
```diff
@@ -44,6 +45,8 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 
+import static org.mockito.Mockito.mock;
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/ResourceManagerTest.java]: unused

File: flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingRestfulGateway.java [hunk 13]
```diff
@@ -281,6 +283,11 @@ public Builder setRequestClusterOverviewSupplier(Supplier<CompletableFuture<Clus
 			return this;
 		}
 
+		public Builder setRequestOverviewForAllJobsSupplier(Supplier<CompletableFuture<JobsOverview>> requestOverviewForAllJobsSupplier) {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/webmonitor/TestingRestfulGateway.java]: We should be able to remove this now.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: yes, my concern was primarily over the long-term. This also becomes relevant if new metrics are added, which will follow a similar pattern as existing ones that don't worry about concurrency.

However, @tillrohrmann's solution isn't even overkill enough, as it could result in exceptions being thrown by metrics (in case of a timeout) which generally shouldn't happen. So we'd have to cache the previous result and return that in case of a timeout.

[the remark this answers is not in the document]
Comment (yanghua) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: @zentol So, I should accept this solution?

> So we'd have to cache the previous result and return that in case of a timeout.

[the remark this answers is not in the document]
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: I think this problem might occur at other places as well. Thus, at some point it might be helpful to either add some tooling to do exactly what @zentol proposed or to make the metric system work with asynchronous metrics.

[the remark this answers is not in the document]
Comment (yanghua) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: @tillrohrmann so, Your suggestion is that we will not solve it in this PR, right?

[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java]: Solving this issue would be overkill for this PR, I agree.

**Comment by yanghua** [comment]
@zentol I have already adopted your previous review suggestions. Does it still need other modifications?

**Comment by zentol** [comment]
I believe you haven't addressed my comments regarding the  `JobManagerMetricsITCase`.

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+import static org.junit.Assert.fail;
+
+/**
+ * Integration tests for proper initialization of the job manager metrics.
+ */
+public class JobManagerMetricsITCase {
+
+	private static final String JOB_MANAGER_METRICS_PREFIX = "localhost.jobmanager.";
+	private static final String TASK_SLOTS_AVAILABLE = "taskSlotsAvailable";
+	private static final String TASK_SLOTS_TOTAL = "taskSlotsTotal";
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: use constants in `MetricNames` instead

**Comment by zentol** [comment]
oh wait you did, Ill take another look.

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	@Before
+	public void setUp() throws Exception {
+		Assume.assumeTrue(
+			"ClusterClient is not an instance of MiniClusterClient",
+			MINI_CLUSTER_RESOURCE.getClusterClient() instanceof MiniClusterClient);
+
+		clusterClient = (MiniClusterClient) MINI_CLUSTER_RESOURCE.getClusterClient();
+		clusterClient.setDetached(true);
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: you can put this job ad an `execute()` call into a separate thread and wait in the main test thread. then you don't have to create a jobgraph or use the cluster client.

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+
+				@Override
+				public void cancel() {
+					sync.releaseBlocker();
+				}
+
+			}).addSink(new PrintSinkFunction());
+
+			try {
+				env.execute();
```
Comment (yanghua) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: @zentol what about this implementation?

File: flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+			} catch (Exception e) {
+				if (!(e instanceof JobExecutionException)) {
+					fail("Execution job failed, reason : " + e.getMessage());
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java]: This won't have the effect you probably intended, as it will not fail the test. You can use a Flink `CheckedThread` to achieve this, and `sync()` at the end of the test.

**Comment by yanghua** [comment]
@zentol thanks, updated.

**Comment by tillrohrmann** [comment]
There is a merge conflict with `ResourceManager`.

**Comment by yanghua** [comment]
@tillrohrmann fixed the conflicts.

**Comment by yanghua** [comment]
I saw some errors not related to this PR change, from `flink-s3-fs-hadoop`. cc @tillrohrmann  and @zentol

**Comment by tillrohrmann** [comment]
There is currently a problem with the `HadoopS3RecoverableWriterExceptionTest`. @kl0u is fixing it. Once this is done, please rebase onto the latest master.

**Comment by yanghua** [comment]
@tillrohrmann rebased and all checks have passed.

**Comment by zentol** [comment]
Ill merge this on monday and start looking into backporting this to 1.5 and 1.6.
